rfcs/014-pagination.md

```diff
@@ -0,0 +1,181 @@
+
+## Terminology
+
+- `bookmark` - is opaque token which would contain information needed to retrieve
+  bookmarked page. The format of the token value MUST NOT be relied upon in the client.
+
+---
+
+# Detailed Description
+
+## Implementation proposal
+
+1) Add new optional query field called `bookmark` to following endpoints
+  - {db}/_all_docs
+  - {db}/_all_docs/queries
+  - _dbs_info
+  - {db}/_design/{ddoc}/_view/{view}
+  - {db}/_design/{ddoc}/_view/{view}/queries
+
+2) Add new optional query field called `page_size` if it is set we would use
+  paginated endpoint otherwise use old code path
+3) Implement per-endpoint configurable max limits to restrict page size of
+  paginated responses
+
+ ```
+ [request_limits]
+ _all_docs = 5000
+ _all_docs/queries = 5000
+ _all_dbs = 5000
+ _dbs_info = 5000
+ _view = 2500
+ _view/queries = 2500
+ _find = 2500
+ ```
+4) Add following additional fields into response:
+  ```
+    "first": "/myddb/_all_docs?bookmark=12345678945621321689",
+    "previous": "/myddb/_all_docs?bookmark=983uiwfjkdsdf",
+    "next": "/myddb/_all_docs?bookmark=12343tyekf3"
+  ```
+
+## Semantics of the implementation
+
+- The bookmark would include information needed to ensure proper pagination
+  without the need to repeat initial parameters of the request.
+- Don't use delayed responses when `bookmark` field is provided
+- Don't use delayed responses when `page_size` query key is specified and when
+  it is below the max limit
+- Return 400 when `bookmark` field is set and other query fields are present
+- Return 400 when `page_size` query key is specified and it is greater than
+  the max limit
+- Return 400 when we stream rows (`page_size` query key wasn't specified) and
+  reach max limit
+- The `previous`/`next`/`first` keys are optional and we omit them for the cases
+  they don't make sense
+- Set default value for `page_size` to `limit` if provided `limit` is less than
+  value configured in `request_limit` of `default.ini` for the given endpoint
+- Set default value for `page_size` to the value configured in `request_limit`
+  of `default.ini` for the given endpoint
+- Once the `limit` is reached the final response will not have a "next" bookmark
+- Once the underlying call to FoundationDB returns less than `page_size`
+  the response will not have a "next" bookmark
+
+## Configuration
+
+The page size limits are configured in `default.ini` (or another ini file) in
+`request_limit` section as follows:
+
+```
+ [request_limits]
+ _all_docs = 5000
+ _all_docs/queries = 5000
+ _all_dbs = 5000
+ _dbs_info = 5000
+ _view = 2500
+ _view/queries = 2500
+ _find = 2500
+```
+
+## Roadmap
+
+- initial implementation as described in this document
+- create API versioning proposal and implement the feature
+- create separate proposal for _changes endpoint
+- implement pagination enabled version of `_all_dbs` which would change response
+  type to be object (using versioned API feature)
+
```

Comment:
Add a not that we would need to converge `_find` endpoint to the same pagination API (using versioned API feature).